File: Chagok/frontend/src/lib/api/consultation.ts
/**
 * Consultation API Client
 * Issue #399: 상담내역 API
 */

import { apiRequest, ApiResponse } from './client';
import type {
  Consultation,
  ConsultationCreate,
  ConsultationUpdate,
  ConsultationListResponse,
  LinkEvidenceRequest,
  LinkEvidenceResponse,
} from '@/types/consultation';

/**
 * Create a new consultation
 */
export async function createConsultation(
  caseId: string,
  data: ConsultationCreate
): Promise<ApiResponse<Consultation>> {
  return apiRequest<Consultation>(`/cases/${caseId}/consultations`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
}

/**
 * Get list of consultations for a case
 */
export async function getConsultations(
  caseId: string,
  params?: { limit?: number; offset?: number }
): Promise<ApiResponse<ConsultationListResponse>> {
  const searchParams = new URLSearchParams();
  if (params?.limit) searchParams.append('limit', params.limit.toString());
  if (params?.offset) searchParams.append('offset', params.offset.toString());

  const queryString = searchParams.toString();
  const url = `/cases/${caseId}/consultations${queryString ? `?${queryString}` : ''}`;

  return apiRequest<ConsultationListResponse>(url, {
    method: 'GET',
  });
}

/**
 * Get a single consultation by ID
 */
export async function getConsultation(
  caseId: string,
  consultationId: string
): Promise<ApiResponse<Consultation>> {
  return apiRequest<Consultation>(`/cases/${caseId}/consultations/${consultationId}`, {
    method: 'GET',
  });
}

/**
 * Update a consultation
 */
export async function updateConsultation(
  caseId: string,
  consultationId: string,
  data: ConsultationUpdate
): Promise<ApiResponse<Consultation>> {
  return apiRequest<Consultation>(`/cases/${caseId}/consultations/${consultationId}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
}

/**
 * Delete a consultation
 */
export async function deleteConsultation(
  caseId: string,
  consultationId: string
): Promise<ApiResponse<void>> {
  return apiRequest<void>(`/cases/${caseId}/consultations/${consultationId}`, {
    method: 'DELETE',
  });
}

/**
 * Link evidence to a consultation
 */
export async function linkEvidence(
  caseId: string,
  consultationId: string,
  evidenceIds: string[]
): Promise<ApiResponse<LinkEvidenceResponse>> {
  const data: LinkEvidenceRequest = { evidence_ids: evidenceIds };

  return apiRequest<LinkEvidenceResponse>(
    `/cases/${caseId}/consultations/${consultationId}/evidence`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    }
  );
}

/**
 * Unlink evidence from a consultation
 */
export async function unlinkEvidence(
  caseId: string,
  consultationId: string,
  evidenceId: string
): Promise<ApiResponse<void>> {
  return apiRequest<void>(
    `/cases/${caseId}/consultations/${consultationId}/evidence/${evidenceId}`,
    {
      method: 'DELETE',
    }
  );
}

File: Chagok/backend/app/api/consultation.py
"""
Consultation API Router - REST endpoints for consultation management
Issue #399: 상담내역 API
"""

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.dependencies import (
    get_db,
    get_current_user_id,
    verify_case_read_access,
    verify_case_write_access,
    get_vector_db_service,
)
from app.db.schemas.consultation import (
    ConsultationCreate,
    ConsultationUpdate,
    ConsultationOut,
    ConsultationListResponse,
    LinkEvidenceRequest,
    LinkEvidenceResponse,
)
from app.services.consultation_service import ConsultationService
from app.domain.ports.vector_db_port import VectorDBPort

router = APIRouter(
    prefix="/cases/{case_id}/consultations",
    tags=["consultations"]
)


@router.post(
    "",
    response_model=ConsultationOut,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new consultation"
)
async def create_consultation(
    case_id: str,
    data: ConsultationCreate,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Create a new consultation record for a case.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    return service.create_consultation(case_id, data, user_id)


@router.get(
    "",
    response_model=ConsultationListResponse,
    summary="List consultations for a case"
)
async def list_consultations(
    case_id: str,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Get all consultations for a case, ordered by date descending.

    Requires read access to the case.
    """
    verify_case_read_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    return service.list_consultations(case_id, user_id, limit=limit, offset=offset)


@router.get(
    "/{consultation_id}",
    response_model=ConsultationOut,
    summary="Get a specific consultation"
)
async def get_consultation(
    case_id: str,
    consultation_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Get a specific consultation by ID.

    Requires read access to the case.
    """
    verify_case_read_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    return service.get_consultation(consultation_id, user_id)


@router.put(
    "/{consultation_id}",
    response_model=ConsultationOut,
    summary="Update a consultation"
)
async def update_consultation(
    case_id: str,
    consultation_id: str,
    data: ConsultationUpdate,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Update a consultation record.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    return service.update_consultation(consultation_id, data, user_id)


@router.delete(
    "/{consultation_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a consultation"
)
async def delete_consultation(
    case_id: str,
    consultation_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Delete a consultation record.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    service.delete_consultation(consultation_id, user_id)


# ============================================
# Evidence Link Endpoints
# ============================================
@router.post(
    "/{consultation_id}/evidence",
    response_model=LinkEvidenceResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Link evidence to a consultation"
)
async def link_evidence(
    case_id: str,
    consultation_id: str,
    data: LinkEvidenceRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Link one or more evidence items to a consultation.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    return service.link_evidence(consultation_id, data, user_id)


@router.delete(
    "/{consultation_id}/evidence/{evidence_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Unlink evidence from a consultation"
)
async def unlink_evidence(
    case_id: str,
    consultation_id: str,
    evidence_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    vector_db_port: VectorDBPort = Depends(get_vector_db_service)
):
    """
    Unlink an evidence item from a consultation.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = ConsultationService(db, vector_db_port=vector_db_port)
    service.unlink_evidence(consultation_id, evidence_id, user_id)
